From 67cf84882660c1c225154028de628538d4341d76 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 24 Mar 2017 17:39:54 -0700
Subject: [PATCH] joyent/node-cueball#106 want INFO logs when backends are
 shuffled into the top N for pool/set Reviewed by: David Pacheco
 <dap@joyent.com>

---
 CHANGES.adoc    | 10 ++++++++++
 lib/pool.js     | 11 +++++++++++
 lib/resolver.js |  5 +++++
 lib/set.js      |  4 ++++
 package.json    |  2 +-
 5 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/CHANGES.adoc b/CHANGES.adoc
index 1aa5dd5..758db06 100644
--- a/CHANGES.adoc
+++ b/CHANGES.adoc
@@ -6,6 +6,16 @@ toc::[]
 
 ## v2.x
 
+### v2.2.4
+
+Maintenance release.
+
+Bugs fixed:
+
+ - #106 want INFO logs when backends are shuffled into the top N for
+   pool/set, and when resolver adds/removes records (only after first
+   successful lookup)
+
 ### v2.2.3
 
 Maintenance release.
diff --git a/lib/pool.js b/lib/pool.js
index 528c077..0cf9e57 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -458,6 +458,17 @@ CueBallConnectionPool.prototype.reshuffle = function () {
 		return;
 	var taken = this.p_keys.pop();
 	var idx = Math.floor(Math.random() * (this.p_keys.length + 1));
+
+	var self = this;
+	var conns = 0;
+	Object.keys(this.p_connections).forEach(function (k) {
+		conns += self.p_connections[k].length;
+	});
+	if (this.p_keys.length > conns && idx < conns) {
+		this.p_log.info('random shuffle puts backend "%s" at idx %d',
+		    taken, idx);
+	}
+
 	this.p_keys.splice(idx, 0, taken);
 	this.rebalance();
 };
diff --git a/lib/resolver.js b/lib/resolver.js
index 09e0e6d..a7f456d 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -861,6 +861,11 @@ CueBallDNSResolver.prototype.state_process = function (S) {
 
 	this.r_backends = newBackends;
 
+	if (oldKeys.length !== 0 && (removed.length > 0 || added.length > 0)) {
+		this.r_log.info({ added: added, removed: removed },
+		    'records changed in DNS');
+	}
+
 	removed.forEach(function (k) {
 		self.r_log.trace('host removed: %s', k);
 		self.emit('removed', k);
diff --git a/lib/set.js b/lib/set.js
index 4775484..5e80905 100644
--- a/lib/set.js
+++ b/lib/set.js
@@ -327,6 +327,10 @@ CueBallConnectionSet.prototype.reshuffle = function () {
 		return;
 	var taken = this.cs_keys.pop();
 	var idx = Math.floor(Math.random() * (this.cs_keys.length + 1));
+	if (this.cs_keys.length > this.cs_target && idx < this.cs_target) {
+		this.cs_log.info('random shuffle puts backend "%s" at idx %d',
+		    taken, idx);
+	}
 	this.cs_keys.splice(idx, 0, taken);
 	this.rebalance();
 };
diff --git a/package.json b/package.json
index 905af66..1adfa25 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "cueball",
-  "version": "2.2.3",
+  "version": "2.2.4",
   "description": "manage a pool of connections to a multi-node service where nodes are listed in DNS",
   "main": "lib/index.js",
   "dependencies": {
-- 
2.21.0

